### PR TITLE
feat(xamarin support): Run stryker.net on xamarin projects

### LINKED ISSUE #572: Compilation error when running against a Xamarin project with a Views Namespace
Running Stryker.NET fails with a compilation error instead of listing mutants that were killed/survived. The same solution compiles fine in VS.

**Steps to reproduce:**
1. Clone [Example project](https://github.com/Project4MobileDev/Stryker.Net.Xamarin.Views.Namespace.Issue)
2. Go to `[REPO]/Source/Example/Module/Tests`
3. Run `dotnet stryker`

```
   _____ _              _               _   _ ______ _______
  / ____| |            | |             | \ | |  ____|__   __|
 | (___ | |_ _ __ _   _| | _____ _ __  |  \| | |__     | |
  \___ \| __| '__| | | | |/ / _ \ '__| | . ` |  __|    | |
  ____) | |_| |  | |_| |   <  __/ |    | |\  | |____   | |
 |_____/ \__|_|   \__, |_|\_\___|_| (_)|_| \_|______|  |_|
                   __/ |
                  |___/


Version 0.11.0 (beta)

[10:37:05 INF] Using D:\Repository\Stryker.Net.Xamarin.Views.Namespace.Issue\Source\Example\Module\Tests\Example.Module.UnitTests.csproj as project file
[10:37:16 INF] File D:\Repository\Stryker.Net.Xamarin.Views.Namespace.Issue\Source\Example\Module\UI\ExampleModule.cs will be excluded from mutation.
[10:37:16 INF] Started initial build using dotnet build
[10:37:19 INF] Initial build successful
[10:37:19 INF] Using testrunner DotnetTest
[10:37:19 INF] Initial testrun started
[10:37:21 INF] Total number of tests found in initial test run: 0
[10:37:21 INF] Initial testrun successful in 2419 ms
[10:37:21 INF] Using 13628 ms as testrun timeout
[10:37:23 WRN] Original project under test D:\Repository\Stryker.Net.Xamarin.Views.Namespace.Issue\Source\Example\Module\UI\bin\Debug\netstandard2.0\Example.Module.dll could not be loaded.
Embedded Resources might be missing.
Mono.Cecil.AssemblyResolutionException: Failed to resolve assembly: 'Xamarin.Forms.Xaml, Version=2.0.0.0, Culture=neutral, PublicKeyToken=null'
   at Mono.Cecil.BaseAssemblyResolver.Resolve(AssemblyNameReference name, ReaderParameters parameters)
   at Mono.Cecil.DefaultAssemblyResolver.Resolve(AssemblyNameReference name)
   at Mono.Cecil.MetadataResolver.Resolve(TypeReference type)
   at Mono.Cecil.Mixin.CheckedResolve(TypeReference self)
   at Mono.Cecil.SignatureReader.ReadCustomAttributeEnum(TypeReference enum_type)
   at Mono.Cecil.SignatureReader.ReadCustomAttributeElement(TypeReference type)
   at Mono.Cecil.SignatureReader.ReadCustomAttributeConstructorArguments(CustomAttribute attribute, Collection`1 parameters)
   at Mono.Cecil.MetadataReader.ReadCustomAttributeSignature(CustomAttribute attribute)
   at Mono.Cecil.CustomAttribute.<Resolve>b__35_0(CustomAttribute attribute, MetadataReader reader)
   at Mono.Cecil.ModuleDefinition.Read[TItem](TItem item, Action`2 read)
   at Mono.Cecil.CustomAttribute.get_ConstructorArguments()
   at Mono.Cecil.ImmediateModuleReader.ReadCustomAttributes(ICustomAttributeProvider provider)
   at Mono.Cecil.ImmediateModuleReader.ReadModule(ModuleDefinition module, Boolean resolve_attributes)
   at Mono.Cecil.ModuleDefinition.Read[TItem](TItem item, Action`2 read)
   at Mono.Cecil.ModuleReader.CreateModule(Image image, ReaderParameters parameters)
   at Mono.Cecil.ModuleDefinition.ReadModule(String fileName, ReaderParameters parameters)
   at Stryker.Core.Initialisation.EmbeddedResourcesGenerator.LoadModule(String assemblyPath, ILogger logger) in D:\a\1\s\src\Stryker.Core\Stryker.Core\Initialisation\EmbeddedResourcesGenerator.cs:line 42
[10:37:23 WRN] Stryker.NET encountered an compile error in D:\Repository\Stryker.Net.Xamarin.Views.Namespace.Issue\Source\Example\Module\UI\ExampleModule.cs with message: The type or namespace name 'Views' does not exist in the namespace 'Example.Module' (are you missing an
 assembly reference?) (Source code: Views)
[10:37:23 WRN] Safe Mode! Stryker will try to continue by rolling back all mutations in method. This should not happen, please report this as an issue on github with the previous error message.
[10:37:23 FTL] Stryker.NET could not compile the project after mutation. This is probably an error for Stryker.NET and not your project. Please report this issue on github with the previous error message.
[10:37:23 ERR] An error occurred during the mutation test run
System.ApplicationException: Internal error due to compile error.
   at Stryker.Core.Compiling.RollbackProcess.RemoveMutantIfStatements(SyntaxTree originalTree, ICollection`1 diagnosticInfo, Boolean devMode) in D:\a\1\s\src\Stryker.Core\Stryker.Core\Compiling\RollbackProcess.cs:line 191
   at Stryker.Core.Compiling.RollbackProcess.Start(CSharpCompilation compiler, ImmutableArray`1 diagnostics, Boolean devMode) in D:\a\1\s\src\Stryker.Core\Stryker.Core\Compiling\RollbackProcess.cs:line 61
   at Stryker.Core.Compiling.CompilingProcess.TryCompilation(MemoryStream ms, CSharpCompilation compilation, EmitResult previousEmitResult, Boolean devMode, Int32 retryCount) in D:\a\1\s\src\Stryker.Core\Stryker.Core\Compiling\CompilingProcess.cs:line 108
   at Stryker.Core.Compiling.CompilingProcess.Compile(IEnumerable`1 syntaxTrees, MemoryStream ms, Boolean devMode) in D:\a\1\s\src\Stryker.Core\Stryker.Core\Compiling\CompilingProcess.cs:line 76
   at Stryker.Core.MutationTest.MutationTestProcess.Mutate(StrykerOptions options) in D:\a\1\s\src\Stryker.Core\Stryker.Core\MutationTest\MutationTestProcess.cs:line 91
   at Stryker.Core.StrykerRunner.RunMutationTest(StrykerOptions options) in D:\a\1\s\src\Stryker.Core\Stryker.Core\StrykerRunner.cs:line 77
[10:37:23 INF] Time Elapsed 00:00:17.6031315
```

**Expected behavior**
Striker should compile the project and list mutants that were killed or survived.

 - OS: Windows 10
 - Type of project: Xamarin Forms v3.6.0.344457
 - Framework Version: core 2.2 (Example.Module.UnitTests), Standard 2.0 (Example.Module)
 - Stryker Version: 0.11.0 (beta)
```

### LINKED ISSUE #756: Mutated assembly is missing embedded resources
**Describe the bug**
Mono.Cecil often fails to load the original assembly which means embedded resources rae missing from mutated assembly.

**Expected behavior**
Mono.Cecil loads the unmutated assembly in order to extract the embedded resources contained.

**Additional context**
Seems to sometimes be caused by file locks on the unmutated assemblies.

### PR DESCRIPTION
**Pull request by felipebaltazar**
fixes #572
Fix #756 

I've created a way to include nuget packages folder for assembly resolver, when project will be compilated
I removed the rule that skips `.xaml.cs` files because it can impact UWP/WPF projects.

Original PR => https://github.com/stryker-mutator/stryker-net/pull/1716

## Demo

image

### CODE REVIEW

**Comment by rouke-broersma** [comment]
It looks like there are changes left over from your previous implementation that are no longer needed. I am trying to understand how the new implementation works before I merge this because you've added it to the assembly reference resolver for embedded resources and I don't really understand how this fixes the problem for xamarin forms projects.

File: src/Stryker.Core/Stryker.Core/Initialisation/CrossPlatformAssemblyResolver.cs [hunk 1]
```diff
@@ -175,7 +183,11 @@ private AssemblyDefinition SearchTrustedPlatformAssemblies(AssemblyNameReference
                     var file = Path.Combine(directory, name.Name + extension);
                     if (!File.Exists(file))
                     {
-                        continue;
+                        var innerFiles = Directory.GetFiles(directory, "*"+extension, SearchOption.AllDirectories);
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core/Initialisation/CrossPlatformAssemblyResolver.cs:186]:
```suggestion
                        var innerFiles = Directory.GetFiles(directory, "*" + extension, SearchOption.AllDirectories);
```

File: src/Stryker.Core/Stryker.Core/Initialisation/CrossPlatformAssemblyResolver.cs [hunk 2]
```diff
@@ -175,7 +183,11 @@ private AssemblyDefinition SearchTrustedPlatformAssemblies(AssemblyNameReference
...
+                        file = innerFiles.FirstOrDefault(f => f.EndsWith(name.Name + extension));
+
+                        if(string.IsNullOrWhiteSpace(file) || !File.Exists(file))
```
Comment (richardwerkman) [inline src/Stryker.Core/Stryker.Core/Initialisation/CrossPlatformAssemblyResolver.cs:189]:
```suggestion
                        if (string.IsNullOrWhiteSpace(file) || !File.Exists(file))
```

**Review by richardwerkman — CHANGES_REQUESTED** [review]
To prove this actually works (and keeps working) I would like to see an added integration test that runs stryker on a xamarin forms project. We have a seperate solution for our integration tests in this repo

**Comment by rouke-broersma** [comment]
I am concerned that with this change stryker hangs on linux for some reason. See the linux integration test pipeline.

File: src/Stryker.Core/Stryker.Core/Initialisation/CsharpProjectComponentsBuilder.cs [hunk 3]
```diff
@@ -87,12 +87,6 @@ private CsharpFolderComposite FindProjectFilesUsingBuildalyzer(IAnalyzerResult a
 
             foreach (var sourceFile in analyzerResult.SourceFiles)
             {
-                // Skip xamarin UI generated files
-                if (sourceFile.EndsWith(".xaml.cs"))
```
Comment (rouke-broersma) [inline src/Stryker.Core/Stryker.Core/Initialisation/CsharpProjectComponentsBuilder.cs:91]: Should we actually do anything with these files? Are they not auto generated?

**Comment by rouke-broersma** [comment]
@felipebaltazar Through investigating your change I found someone giving a hint for a better solution. Since we actually already know the location of every single referenced assembly we do not actually have to search the filesystem for assemblies at all. I have replaced our AssemblyResolver with an implementation that literally loads the assemblies from disk using a list of filepaths, this solution also loads the xamarin.forms dll and causes Mono.Cecil to be able to load the embedded resources. I think this might also finally fix #756

File: src/Stryker.Core/Stryker.Core/Initialisation/CsharpProjectComponentsBuilder.cs (hunk 3 above)
Comment (felipebaltazar) [reply src/Stryker.Core/Stryker.Core/Initialisation/CsharpProjectComponentsBuilder.cs:91]: No... only `xaml.g.cs` are auto-generated.. and striker already have a way to work with antogen files

**Comment by felipebaltazar** [comment]
I'm not sure about what is breaking the pipe for integration tests...
But all the tests run with success. Seems like some dependency that was not resolved.

image

**Comment by rouke-broersma** [comment]
The libman error can be ignored, it is a reliability issue with the Javascript package source. We use it to download the sources for the html report.

There is another issue that is an actual problem, I'll try to look at it today.

**Comment by rouke-broersma** [comment]
Something is wrong with this implementation for .NET framework projects. I have asked the assembly resolver author for assistance: https://github.com/atykhyy/cecil-msbuild-helper/issues/1

**Comment by atykhyy** [comment]
The discussion at atykhyy/cecil-msbuild-helper#1 drifted away from the helper itself, so I follow up here.

> Cecil in its default configuration for some reason cannot open assemblies using xamarin due to missing the xamarin core assembly

If you only need Cecil to extract embedded resources, why read assembly references at all? Have you tried using Cecil's deferred reading mode?

**Comment by rouke-broersma** [comment]
> The discussion at [atykhyy/cecil-msbuild-helper#1](https://github.com/atykhyy/cecil-msbuild-helper/issues/1) drifted away from the helper itself, so I follow up here.
> 
> > Cecil in its default configuration for some reason cannot open assemblies using xamarin due to missing the xamarin core assembly
> 
> If you only need Cecil to extract embedded resources, why read assembly references at all? Have you tried using Cecil's deferred reading mode?

We are using Immediate reading mode. I think this was because otherwise the resourcestream to the embedded resources would be disposed by the time we got around to compiling.

I must admit that I know very little about Cecil and Assembly loading so perhaps I'm simply doing something horribly wrong here.

We load the embedded resources in this way during our initialization phase: 
https://github.com/stryker-mutator/stryker-net/blob/18967c2944cafda6a8367f6248e60e233aebad3a/src/Stryker.Core/Stryker.Core/Initialisation/EmbeddedResourcesGenerator.cs

We use the embedded resource stream here during the compilation phase, which is after the mutation phase (creating bugs in the roslyn syntaxtrees): 
https://github.com/stryker-mutator/stryker-net/blob/18967c2944cafda6a8367f6248e60e233aebad3a/src/Stryker.Core/Stryker.Core/Compiling/CsharpCompilingProcess.cs#L149

through here -> https://github.com/stryker-mutator/stryker-net/blob/18967c2944cafda6a8367f6248e60e233aebad3a/src/Stryker.Core/Stryker.Core/Initialisation/Buildalyzer/IAnalyzerResultExtensions.cs#L30-L33

**Comment by rouke-broersma** [comment]
> Cecil's handling of disposal is certainly confusing. I've been bitten by it myself. However, in `ReadResourceDescriptionsFromModule`, which is called before the module is disposed, your code copies resource streams obtained from Cecil into newly created memory streams, so it should not be affected.

I will try Deferred mode and see if that solves our problems, thank you!

**Comment by rouke-broersma** [comment]
Thank you so much for taking the time to look into this @felipebaltazar! I implemented the suggestion made by @atykhyy in #1816 and it appears to solve the problems in my test project. If that also works for your projects then sadly I think this PR is no longer necessary however the new solution would be much simpler and without your PR we would not have gotten there so thank you!

**Comment by felipebaltazar** [comment]
> Thank you so much for taking the time to look into this @felipebaltazar! I implemented the suggestion made by @atykhyy in #1816 and it appears to solve the problems in my test project. If that also works for your projects then sadly I think this PR is no longer necessary however the new solution would be much simpler and without your PR we would not have gotten there so thank you!

@rouke-broersma that's awesome! Seems like the best solution that we can do.. I will test the branch with this change for some projects that I've used xamarin and refit.. I will bring some feedback to you soon

Thanks for your attention 
Keep rocking 👊

**Pull request closed without merging**